### ISSUE TITLE
Incorrect warning about Objective-C Category method not being documented

### ISSUE THREAD

**Issue by smokris** [doxygen/doxygen#10897]
**Describe the bug**

For a documentation block on an Objective-C Category method, Doxygen emits this warning:

> GraphicsWindowDrag.h:1: warning: Member initDrag() (function) of class GraphicsWindow is not documented.

**To Reproduce**

- extract this archive: [objctest.zip](https://github.com/doxygen/doxygen/files/15444627/objctest.zip)
- change into that folder and run `doxygen .`
- Doxygen emits the above warning

**Expected behavior**

- since the method is properly documented, Doxygen should not emit a warning

**Version**

- Doxygen 1.11.0, macOS, Homebrew
- The bug did not happen in Doxygen 1.10.0 or earlier

**Comment by albert-github** [comment]
Bisecting gives:
```
b9cb1df78102b0390488082d93283c0aa49b96b6 is the first bad commit
commit b9cb1df78102b0390488082d93283c0aa49b96b6

Date:   Fri Apr 12 21:29:33 2024 +0200

    Refactoring: apply rule of 0 or rule of 5 for special members (part 2)
```

In the example I see:
```
@interface GraphicsWindow : NSWindow
@end

@interface GraphicsWindow (Drag)
- (void)initDrag;
@end

@implementation GraphicsWindow (Drag)
/**
 * some documentation
 */
- (void)initDrag
{
}
@end
```

I'm not really familiar with Object-C but it looks like there are 2 interfaces and also doxygen thinks these are 2 interfaces:
![image](https://github.com/doxygen/doxygen/assets/5223533/b76eace0-bacd-4b07-bf4f-60586714ac9e)

and thus there is no relation between the `initDrag` of `GraphicsWindow` and `initDrag` of `GraphicsWindow (Drag)`  and in this sense the warning was missing in the 1.10.0 version though doxygen 1.10.0 thinks theer is a relation as it says:
![image](https://github.com/doxygen/doxygen/assets/5223533/8fe6a20d-723e-4bbd-af49-54ab82fe0aa3)


As said "I'm not really familiar with Object-C" so @smokris maybe you can enlighten me a bit.

**Comment by doxygen** [comment]
@smokris Please verify if the referenced commit fixes the problem for you. Do not close the issue, this will be done automatically when the next official release becomes available.

**Comment by albert-github** [comment]
@doxygen 

On the page interface_graphics_window.html we see:
![image](https://github.com/doxygen/doxygen/assets/5223533/bfd48279-092d-4040-bda7-eea76f6d226b)

Though the link `GraphicsWindow(Drag)` pints to interface_graphics_window.html#a2b2ecc38c5eb5e592dfc18b616063236).
and I would have expected a link to the page category_graphics_window_07_drag_08.html 
(or is this also a lack in my knowledge?)

**Comment by doxygen** [comment]
@albert-github You are right indeed. This was already wrong in 1.10.0 however, but should be fixed now.

**Comment by albert-github** [comment]
@doxygen This problem looks like to be solved thoug I added a small reference page:
```
/** \page pg1 reference page

Reference to `GraphicsWindow::initDrag`:
- \ref GraphicsWindow::initDrag

Reference to `GraphicsWindow(Drag)::initDrag`:
- \ref GraphicsWindow(Drag)::initDrag
*/
```
and this results in:
![image](https://github.com/doxygen/doxygen/assets/5223533/d3fef713-7ffc-4a92-b33c-2618fa54b9d7)

The first text line with the part between the backticks and the first `\ref` look OK and have correct references.

The `` Reference to `GraphicsWindow(Drag)::initDrag`:`` has a strange output result and also the reference is to the page interface_graphics_window.html and not to the function I would have expected: category_graphics_window_07_drag_08.html#a2b2ecc38c5eb5e592dfc18b616063236

Also the `\ref` is strange and the link is not as I would have expected.

My example: [example.tar.gz](https://github.com/doxygen/doxygen/files/15448088/example.tar.gz)

**Comment by albert-github** [comment]
Probably something with the references for Objective-C doesn't work properly as also an example like:
```
@interface GraphicsWindow : NSWindow
@end

@interface GraphicsWindow (Drag)
- (void)initDrag;
@end

@implementation GraphicsWindow (Drag)
- (void)initDrag
{
}
@end


/**
 * @fn GraphicsWindow(Drag)::initDrag
 * some documentation
 */
```
gives problems as it cannot find the `@fn GraphicsWindow(Drag)::initDrag` to be referenced

**Comment by doxygen** [comment]
@albert-github The 'problem' is that `@interface X(Y)` is an extension of `@interface X`, so a user would still refer to a method `init` added by extension `X(Y)` as `X::init` and not as `X(Y)::init`. This is also why doxygen adds `init` to `X`.

**Comment by albert-github** [comment]
Still I don't get it (probably my lack of knowledge though).

It is strange to me that the part:
```
Reference to `GraphicsWindow(Drag)::initDrag`:
- \ref GraphicsWindow(Drag)::initDrag
```
gives 2 different outputs.

Also how to document the component `GraphicsWindow(Drag)` and parts in it properly?

When having:
```
@interface GraphicsWindow : NSWindow
@end

@interface GraphicsWindow (Drag)
- (void)initDrag;
@end

@implementation GraphicsWindow (Drag)
- (void)initDrag
{
}
@end

/**
 * @interface GraphicsWindow
 * the initial interface
 */

/**
 * @interface GraphicsWindow(Drag)
 * the Drag interface
 */

/**
 * @fn GraphicsWindow::initDrag
 * some documentation
 */
```


I get:
```
.../objctest/GraphicsWindowDrag.h:20: warning: the name 'Drag' supplied as the argument of the \class, \struct, \union, or \include command is not an input file
.../objctest/GraphicsWindowDrag.h:4: warning: Compound GraphicsWindow(Drag) is not documented.
.../objctest/GraphicsWindowDrag.h:25: warning: no matching class member found for
  GraphicsWindow::initDrag
Possible candidates:
  'void GraphicsWindow(Drag)::initDrag()' at line 1 of file D:/doxygen_bugs/_patches/issues/issues_10/issue_10897/ex_3/objctest/GraphicsWindowDrag.h
.../objctest/GraphicsWindowDrag.h:1: warning: Member initDrag() (function) of class GraphicsWindow is not documented.
.../objctest/GraphicsWindowDrag.h:1: warning: Member initDrag() (function) of category GraphicsWindow(Drag) is not documented.
```

Example 3: [example_3.tar.gz](https://github.com/doxygen/doxygen/files/15467120/example_3.tar.gz)

**Comment by smokris** [comment]
Thanks for your work on this, @albert-github and @doxygen.

> Please verify if the referenced commit fixes the problem for you

✅ It does.  I made a local build at 05ef427, and verified that the original problem I described above is fixed — it no longer emits a warning, and it correctly associates the Objective-C Category with the existing class that it modifies.

> As said "I'm not really familiar with Object-C" so \@smokris maybe you can enlighten me a bit.

Objective-C Categories are commonly used if you have a large/complex class, and you want to logically divide the methods into multiple groups, in order to (1) organize how the public API is presented in documentation and (2) split the implementation into multiple source files.

This is similar to Doxygen's [Member Groups](https://doxygen.nl/manual/grouping.html#memgroup) feature — maybe Doxygen should automatically place each Objective-C Category's methods in a Member Group when displaying them on the main class's page?

For example, in this snippet, I've manually inserted a Doxygen Member Group around each Objective-C Category's methods, but I think the expectation among Objective-C developers is that the documentation tool would perform this grouping automatically:
```objc
@interface GraphicsWindow : NSWindow
- (void)show;
@end
@implementation GraphicsWindow
/// Makes the window visible and focused.
- (void)show { }
@end

@interface GraphicsWindow (Drag)
/// @name Drag
///@{
- (BOOL)acceptsDrag:(NSEvent *)e;
- (void)performDrag:(NSEvent *)e;
///@}
@end
@implementation GraphicsWindow (Drag)
/// Returns YES if the window is willing to accept the specific drag event.
- (BOOL)acceptsDrag:(NSEvent *)e { return NO; }
/// Processes the drag event.
- (void)performDrag:(NSEvent *)e { }
@end

@interface GraphicsWindow (Keyboard)
/// @name Keyboard
///@{
- (BOOL)acceptsKeypress:(NSEvent *)e;
- (void)performKeypress:(NSEvent *)e;
///@}
@end
@implementation GraphicsWindow (Keyboard)
/// Returns YES if the window is willing to accept the specific keypress.
- (BOOL)acceptsKeypress:(NSEvent *)e { return NO; }
/// Processes the keypress event.
- (void)performKeypress:(NSEvent *)e { }
@end
```

Related: #5378
